server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/LinearFillOperator.java

```diff
@@ -0,0 +1,189 @@
+  private int currentFilledColumnIndex;
+  // indicate whether we can call child.next()
+  // it's used to make sure that child.next() will only be called once in LinearFillOperator.next();
+  private boolean canCallNext;
+  // indicate whether there is more TsBlock for child operator
+  private boolean noMoreTsBlock;
+
+  public LinearFillOperator(
+      OperatorContext operatorContext, LinearFill[] fillArray, Operator child) {
+    this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+    checkArgument(
+        fillArray != null && fillArray.length > 0, "fillArray should not be null or empty");
+    this.fillArray = fillArray;
+    this.child = requireNonNull(child, "child operator is null");
+    this.outputColumnCount = fillArray.length;
+    this.cachedTsBlock = new ArrayList<>();
+    this.nextTsBlockIndex = new int[outputColumnCount];
+    Arrays.fill(this.nextTsBlockIndex, 1);
+    this.cachedFilledValueColumns = new Column[outputColumnCount];
+    this.currentFilledColumnIndex = 0;
+    this.canCallNext = false;
+    this.noMoreTsBlock = true;
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public ListenableFuture<Void> isBlocked() {
+    return child.isBlocked();
+  }
+
+  @Override
+  public TsBlock next() {
+
+    // make sure we call child.next() at most once
+    if (cachedTsBlock.isEmpty()) {
+      canCallNext = false;
+      TsBlock nextBlock = child.next();
+      // child operator's calculation is not finished, so we just return null
+      if (nextBlock == null || nextBlock.isEmpty()) {
+        return nextBlock;
+      } else { // otherwise, we cache it
+        cachedTsBlock.add(nextBlock);
+      }
+    }
+
+    TsBlock block = cachedTsBlock.get(0);
+    long currentEndTime = block.getEndTime();
+    // use cached TsBlock to keep filling remaining column
+    while (currentFilledColumnIndex < outputColumnCount) {
+      // current valueColumn can't be filled using current information
+      if (fillArray[currentFilledColumnIndex].needPrepareForNext(
+          currentEndTime, block.getColumn(currentFilledColumnIndex))) {
+
+        if (canCallNext) { // if we can call child.next(), we call that and cache it in
+          // cachedTsBlock
+          canCallNext = false;
+          TsBlock nextBlock = child.next();
+          // child operator's calculation is not finished, so we just return null
+          if (nextBlock == null || nextBlock.isEmpty()) {
+            return nextBlock;
+          } else { // otherwise, we cache it
+            cachedTsBlock.add(nextBlock);
+          }
+        }
+
+        // next TsBlock has already been in the cachedTsBlock
+        while (nextTsBlockIndex[currentFilledColumnIndex] < cachedTsBlock.size()) {
+          TsBlock nextTsBlock = cachedTsBlock.get(nextTsBlockIndex[currentFilledColumnIndex]);
+          nextTsBlockIndex[currentFilledColumnIndex]++;
```

Comment:
I think you abused the word `next`. So much `next` in your code, `nextTime`, `nextTimeInCurrentColumn`, `nextBlock`, `nextTsBlockIndex`. The key is they DON'T mean exactly the same thing. Look at the code, `nextBlock` just means next Block, but `nextTsBlock` is the same thing? You have to rename some parameters.

Reply:
`nextBlock` is same as `nextTsBlock`, I will rename `nextBlock` to `nextTsBlock`.